### PR TITLE
Fix commit retrieval by tag

### LINKED ISSUE #21687: Incorrect latest commit message in UI when viewing tag
### Description

The latest commit is incorrectly rendered when viewing a tag, the first line (commit title) is absent. It's okay when viewing a branch.

Not OK (tag view) https://try.gitea.io/silverwind/remap-css/src/tag/5.0.4:

image

OK (branch view) https://try.gitea.io/silverwind/remap-css/src/branch/master:

image

### Gitea Version

1.19

### Can you reproduce the bug on the Gitea demo site?

Yes

### PR DESCRIPTION
**Pull request by silverwind**
It is not correct to return tag data when commit data is requested, so remove the hacky code that overwrote parts of a commit with parts of a tag.

This fixes commit retrieval by tag for both the latest commit in the UI and the commit info on tag webhook events.

Fixes: https://github.com/go-gitea/gitea/issues/21687
Replaces: https://github.com/go-gitea/gitea/pull/21693

Screenshot 2022-11-13 at 15 26 37

image

### CODE REVIEW

**Comment by silverwind** [comment]
Lots of tests need alteration because of the new commit introduced in the test repo, working through them 😉

File: modules/git/repo_tag_test.go [hunk 1]
```diff
@@ -26,11 +26,14 @@ func TestRepository_GetTags(t *testing.T) {
 		assert.NoError(t, err)
 		return
 	}
-	assert.Len(t, tags, 1)
+	assert.Len(t, tags, 2)
 	assert.Equal(t, len(tags), total)
-	assert.EqualValues(t, "test", tags[0].Name)
-	assert.EqualValues(t, "3ad28a9149a2864384548f3d17ed7f38014c9e8a", tags[0].ID.String())
+	assert.EqualValues(t, "signed-tag", tags[0].Name)
```
Comment (KN4CK3R) [inline modules/git/repo_tag_test.go:30]: Is this always element 0 now that there are two tags?

Comment (silverwind) [reply modules/git/repo_tag_test.go:31]: Haven't checked the code but generally order depends on `tag.sort` git config.

File: modules/git/repo_commit_test.go [hunk 2]
```diff
@@ -50,7 +50,7 @@ func TestGetTagCommitWithSignature(t *testing.T) {
 	assert.NotNil(t, commit)
 	assert.NotNil(t, commit.Signature)
 	// test that signature is not in message
-	assert.Equal(t, "tag", commit.CommitMessage)
+	assert.Equal(t, "signed-commit\n", commit.CommitMessage)
```
Comment (silverwind) [inline modules/git/repo_commit_test.go:52]: I'm convinced this `\n` is a bug of its own in `CommitFromReader`, but I'm not sure I want to touch that here as well. FTR, commit was generated via:

```
git commit --allow-empty -m "signed-commit"
```

File: modules/git/repo_tag_test.go (hunk 1 above)
Comment (silverwind) [reply modules/git/repo_tag_test.go:30]: It sorts by date so it should be stable ordering:

```
$ GIT_DIR=modules/git/tests/repos/repo1_bare git for-each-ref --sort '-*creatordate' refs/tags
36f97d9a96457e2bab511db30fe2db03893ebc64 tag	refs/tags/signed-tag
3ad28a9149a2864384548f3d17ed7f38014c9e8a tag	refs/tags/test
```

Also, further down it uses `sortTagsByTime` to sort again (probably pointlessly).

File: modules/git/repo_commit_test.go (hunk 2 above)
Comment (silverwind) [reply modules/git/repo_commit_test.go:52]: Pushed a fix for this as well where `CommitFromReader` trims of a single newline from the commit message.

**Comment by zeripath** [comment]
Please note I was not the originator of this behaviour - I simply copied it because it appeared to be relied upon and it would have been too breaking a change at the time I copied the behaviour. 

You're going to have to check everywhere that a tag can be used as commit ref and double check that it's behaving correctly.

**Comment by silverwind** [comment]
Might want to keep it to 1.19 then. Tests alone will not suffice to identify potential issues, so it will need manual testing as well.

Thought I'm convinced this is the right thing to do, both with the commit retrieval and the newline removal from the commit message.

**Comment by zeripath** [comment]
There was another PR recently I think that touched this too wasn't there? @KN4CK3R did you propose/merge that?

We need to really check that this isn't doing the wrong thing.

**Comment by KN4CK3R** [comment]
I had the same changes but when the tests started failing I was unsure if it was the right way. @silverwind going the same way now but with more stamina 😉

**Comment by silverwind** [comment]
Test failures related to the newline removal on commit message are still plagueing me here. I think I may back out of that changes as it it unrelated, strictly speaking (but still a bug).

**Comment by silverwind** [comment]
Newline related changes backed out. If I'm correct, CI should now pass.

**Comment by silverwind** [comment]
CI passes, ready for review.

**Review by KN4CK3R — APPROVED** [review]

**Comment by KN4CK3R** [comment]
Are you planning a new PR with the `\n` fix?

**Comment by silverwind** [comment]
Maybe later, but go ahead if you want to now.

**Comment by silverwind** [comment]
Anyone else willing/able to review? I think this is the right thing to do, and I don't really expect any breakage as commit metadata is still there.

**Comment by silverwind** [comment]
> You're going to have to check everywhere that a tag can be used as commit ref and double check that it's behaving correctly.

Care to elaborate this again? At least as far as git is concerned, both a tag and a commit hash are `tree-ish` objects and can be used interchangably, so I see no risk of this breaking anything.

**Review by delvh — APPROVED** [review]
I certainly see what you mean by `hacky code`.
This rather begs the question: Why was this done in the first place?

**Comment by delvh** [comment]
Should we backport this PR to 1.19 as well, or would that be a too short span to fix potential regressions?

**Comment by silverwind** [comment]
I would not backport for now. Maybe in a few weeks if there's no issues discovered.

**Comment by silverwind** [comment]
Actually, I think I would try this for 1.19, as I would like to have this rather sooner than latter. I wonder if it is enough for gitea bot to create a backport if I add the tag now.

**Comment by silverwind** [comment]
❤️  @GiteaBot

**Pull request merged**